pallets/pallet-did-lookup/src/migrations.rs

```diff
@@ -16,74 +16,144 @@
 	}
+}
+
+pub struct EthereumMigration<T>(PhantomData<T>);
 
+impl<T: crate::pallet::Config> OnRuntimeUpgrade for EthereumMigration<T>
+where
+	T::AccountId: Into<LinkableAccountId>,
+{
 	fn on_runtime_upgrade() -> frame_support::weights::Weight {
-		// Account for the new storage version written below.
-		let initial_weight = T::DbWeight::get().writes(1);
-
-		// Origin was disabled, so there cannot be any existing links. But we check just
-		// to be sure.
-		let total_weight: Weight =
-			ConnectedDids::<T>::iter().fold(initial_weight, |total_weight, (account, record)| {
-				ConnectedAccounts::<T>::insert(record.did, account, ());
-				// One read for the `ConnectedDids` entry, one write for the new
-				// `ConnectedAccounts` entry.
-				total_weight.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+		if Pallet::<T>::on_chain_storage_version() == Pallet::<T>::current_storage_version() {
+			// already on version 3
+			<T as frame_system::Config>::DbWeight::get().reads_writes(1, 0)
+		} else {
+			log::info!("🔎 DidLookup: Initiating migration");
+			let mut connected_dids = 0u64;
+			let mut connected_accounts = 0u64;
+
+			// Migrate connected DIDs
+			// We should not write to the same storage item during drain because it can lead
+			// to undefined results. Thus, we write to a temporary storage and move that at
+			// the end. Else we iterate over every key more or less twice.
+			ConnectedDids::<T>::drain().for_each(|(acc_id32, value)| {
+				log::debug!("🔎 #{} Migrating ConnectedDid for account", connected_dids,);
+				let acc_id: LinkableAccountId = acc_id32.into();
+				TmpConnectedDids::<T>::insert(acc_id, value);
+				connected_dids = connected_dids.saturating_add(1);
+			});
+			log::info!("🔎 DidLookup: Migrated all {} ConnectedDids", connected_dids);
+
+			// Migrate accounts
+			ConnectedAccounts::<T>::drain().for_each(|(did_id, acc_id32, val)| {
+				log::debug!("🔎 #{:?} Migrating ConnectedAccount", connected_accounts,);
+				let acc_id: LinkableAccountId = acc_id32.into();
+				TmpConnectedAccounts::<T>::insert(did_id, acc_id, val);
+				connected_accounts = connected_accounts.saturating_add(1);
 			});
+			log::info!("🔎 DidLookup: Migrated all {:?} ConnectedAccounts", connected_accounts);
 
-		Pallet::<T>::current_storage_version().put::<Pallet<T>>();
+			// Move TmpStorage
+			move_storage::<Pallet<T>>(b"TmpConnectedDids", b"ConnectedDids");
+			move_storage::<Pallet<T>>(b"TmpConnectedAccounts", b"ConnectedAccounts");
 
+			Pallet::<T>::current_storage_version().put::<Pallet<T>>();
+
+			<T as frame_system::Config>::DbWeight::get().reads_writes(
+				(connected_dids.saturating_add(connected_accounts)).saturating_mul(2),
+				(connected_dids.saturating_add(connected_accounts))
+					.saturating_mul(2)
+					.saturating_add(1),
+			)
+		}
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		assert_eq!(Pallet::<T>::on_chain_storage_version() < Pallet::<T>::current_storage_version());
```

Comment:
I think what you were looking for is

```suggestion
		assert!(Pallet::<T>::on_chain_storage_version() < Pallet::<T>::current_storage_version());
```